### ISSUE TITLE
Create UI for a "detect bugs with similar domain names" feature

### ISSUE THREAD

**Issue by zoepage** [webcompat/webcompat.com#1826]
This is a new approach to #307. 
Before I can started on this, I'd like to clarify the requirements and some questions. 

### Intro:
The idea is to create an UI to find issues with similar domain names and maybe even more. This feature should be seen as "power user feature", so it would be enabled for users, who are logged in.  

For the requests the easiest way for a first version would be to use the Github search API.

### Description:
After adding an URL (min. 3 chars) to the input in our bug report form, a dropdown would show up and suggest related issues. Click an issue to "link" it to the one you are reporting right now. Focus on input (always a min. of 3 chars) gives you again the dropdown, if there are still issues that can be suggested. A selected issue can be deleted (see mock).

Suggested issues:
![img_20170919_105920](https://user-images.githubusercontent.com/562354/30584734-f0af3272-9d2a-11e7-801a-8fcb735ec4aa.jpg)

After selecting issues:
![img_20170919_105914](https://user-images.githubusercontent.com/562354/30584696-d2aa589c-9d2a-11e7-8efc-566acaeebb21.jpg)

List of issues (imagine several of those in an actual list ;) ):
![img_20170919_110058](https://user-images.githubusercontent.com/562354/30584772-11c06292-9d2b-11e7-810b-8d86d2d3fdde.jpg)

Single issue:
![img_20170919_105937](https://user-images.githubusercontent.com/562354/30584800-2a7ed7f0-9d2b-11e7-980f-b8b76a29a53d.jpg)

#### What is the criteria? 
- Do we filter by `URL`?
- Do we treat `www.firefox.org` and `www.firefox.org/subpage` the same way? (Search API does, also includes mentions of a string in the comments into the search results)
- Do we filter just by `title` or `content including comments`? (Please remember we changed the working of the types + the titles are not very explicit as they are autogenerated.)
- Do we filter by `URL and keyword`? If both, how would we include this in the search?

### Concerns / Improvements (after 1st version is done):
- Performance probably will become an issue at some point due to the search API requests. This is why saving the issue list in a local database or in-browser storage would be a good way to go. 
We could also add a crobjob at some point to reference the issues and minimize the needed requests.

- What happens when we hit the max of API requests? (Reset after 24h)

### References:
Older info can be found here #307, #864 and #865.

**Comment by adamopenweb** [comment]
I really like the workflow, as mentioned our issue titles aren't very descriptive compared to Bugzilla. API requests shouldn't be high for those who've clicked the report site issue button, though they may not be signed in depending on their activity here.

Not sure if this is helpful: As a person doing triage, I would like to see what other issues exist for this domain. If there is a report for this domain that has a lot of activity, it's possible we have a common duplicate. This feature is useful to me in that some bugs may have linked reports, but I would prefer to see this list myself when I'm working on the bug. It would be more current and I'm not trusting that the reporter did the job for me.

**Comment by karlcow** [comment]
@zoepage Thanks for restarting the discussion on this.

> This feature should be seen as "power user feature", so it would be enabled for users, who are logged in.

I'm not convinced about this. One of the workflow is specifically to start filing a bug and at the end to be able to report it as anonymous or under your github name. It is specifically true when you test for a specific browser on mobile and/or another browser where you are not logged in usually.

> For the requests the easiest way for a first version would be to use the Github search API.

We can try a prototype, but I have an intuition that it will create issues. Let's explore a bit more.

>The Search API has a custom rate limit. For requests using Basic Authentication, OAuth, or client ID and secret, you can **make up to 30 requests per minute**. For unauthenticated requests, the rate limit allows you to make up to 10 requests per minute.

In the scenario being introduced here:

> After adding an URL (min. 3 chars) to the input in our bug report form, a dropdown would show up and suggest related issues. 

I type 'www', the search API kicks in for all domains which contains 'www' which is the majority. So at this stage not helpful. then you could decide to wait for more characters. Let's say you search every two characters.

* 'www.f' - 2nd hit to search API
* 'www.fir' - 3rd hit to search API
* 'www.firef' - 4th hit to search API 

All of this in a couple of seconds. 30/minute might be enough even 10/minute by minute if we consider that someone is unlikely to report many bugs in one minute. Also we have to take into account that it's an HTTP requests with a couple of RTT. 

For example from Japan right now, for this URI aka all issues page going through webcompat.com
`https://webcompat.com/api/issues?page=1&per_page=50&state=open&stage=all&sort=created&direction=desc` I get `1s`.

![capture d ecran 2017-09-22 a 16 16 37](https://user-images.githubusercontent.com/505230/30733377-aa0971c8-9fb1-11e7-85a4-aff179804f59.png)

and for searching youtube on webcompat.com all issues page. It does this request:

`https://api.github.com/search/issues?page=1&per_page=50&stage=all&sort=created&q=youtube+state%3Aopen+repo%3Awebcompat%2Fweb-bugs&order=desc`

which is `1659ms`

![capture d ecran 2017-09-22 a 16 19 11](https://user-images.githubusercontent.com/505230/30733435-dc273c6c-9fb1-11e7-9ed1-295eb271e606.png)

searching again for `Google`

`https://api.github.com/search/issues?page=1&per_page=50&stage=all&sort=created&q=google+state%3Aopen+repo%3Awebcompat%2Fweb-bugs&order=desc`

![capture d ecran 2017-09-22 a 16 21 13](https://user-images.githubusercontent.com/505230/30733486-17d7e644-9fb2-11e7-97d6-207794bc8c75.png)

I wonder how we could improve this. **TO THINK**.

## Report Site Issue Button

When the issue is reported through the bug reporter they do not type the URL, it comes automagically. So we would have to define what is the meaningful way of handling this. Specifically in context such as facebook.com or vk.com where we have a lot of them.

## GitHub Search API sucks

It has been for a long time. It is useful for certain simple keyword searches but as soon as you get a `.` into there. Things are falling apart, and that's a major issue for 😆 domain names.

Check these two searches: `is:issue facebook.com in:title` and `is:issue facebook in:title` 

facebook.com returns 9 results, facebook returns 299 results with a majority of facebook.com

* https://github.com/webcompat/web-bugs/issues?utf8=%E2%9C%93&q=is%3Aissue%20facebook.com%20in%3Atitle%20
* https://github.com/webcompat/web-bugs/issues?utf8=%E2%9C%93&q=is%3Aissue%20facebook%20in%3Atitle%20

(btw I have the same issue for #60 the domain name feed. :/ )

I don't see much hope without having our local DB of domain names. it will be useful for other stuffs.

## Search criteria
The fuzzy search mechanism becomes a "lot easier" to tweak once we have a reliable way to query the data.

**Comment by zoepage** [comment]
@adamopenweb 

> I would like to see what other issues exist for this domain. 

Makes sense. When you'd get a list displayed with URL's and the type (as it is right now), would this help already? Or would you need to have a way to e.g. open them in a new tab?

@karlcow 

> I'm not convinced about this. One of the workflow is specifically to start filing a bug and at the end to be able to report it as anonymous or under your github name.

As we discussed at the work week on Berlin, we have two types of reporters. The ones, who don't have much time to report and those, who put in as much detail as possible. The only way to split them up right now (especially with taking the extension in account) would be to be logged in, I guess? Maybe you have another idea? So we would keep it simple as possible for the "quick reporters", but give the opportunity to others to link issues or even prevent duplicates.

> I type 'www', the search API kicks in for all domains which contains 'www' which is the majority. So at this stage not helpful. then you could decide to wait for more characters. Let's say you search every two characters.

The shortest possible domain would be some thing like `x.y`, but we could check the first 2 chars for the `www.`string and fire the event starting at the 5th character then.

> I wonder how we could improve this. TO THINK.

With a localDB, performance should get much better as we don't have to go through the GitHubAPI / HTTP requests. But for a prototype, this might be too much?

> When the issue is reported through the bug reporter they do not type the URL, it comes automagically. So we would have to define what is the meaningful way of handling this. Specifically in context such as facebook.com or vk.com where we have a lot of them.

This is why this feature would only be available on the site itself and not in the extension.

> I don't see much hope without having our local DB of domain names. it will be useful for other stuffs.

So you'd dismiss this proposal completely and go with the localDB solution?

**Comment by karlcow** [comment]
( @zoepage thinking out loud, all these discussions are not necessary fully matured. In case it is not clear, I have tendencies to share the stream of my thoughts. Not a final answer. Comments are here to think.)


## About type of reporters

@zoepage said:
>As we discussed at the work week on Berlin, we have two types of reporters. The ones, who don't have much time to report and those, who put in as much detail as possible.

Yes. This is orthogonal to the login status.

@zoepage said:
> The only way to split them up right now (especially with taking the extension in account) would be to be logged in, I guess? 

Not necessary. For example, when I report issues with Firefox Nightly, I'm never logged initially. I start the reporting, and once it is done, at the end of the process, when I submit I choose the option to do it through GitHub but all the work is already done.

![capture d ecran 2017-09-27 a 08 45 21](https://user-images.githubusercontent.com/505230/30889277-48592e66-a360-11e7-87b9-b93fd1c1c565.png)

@zoepage said:
> Maybe you have another idea? So we would keep it simple as possible for the "quick reporters", but give the opportunity to others to link issues or even prevent duplicates.

I have the feeling that it is a powerful feature for both categories of users in a different way. (Note that I assume below that power user != logged in user)

* As a power user, it could help me find, remember the original issue (of many duplicates). if I'm in the process of reporting.
* As a power user, it could help me with triage (!= different from bug reporting form)
* As a new user, it could help me to understand if I should report this issue or not. (Maybe it's an outcome we don't want.)

Maybe we have forgotten of what are the outcomes. Let's forget about power/new users for a minute.

We search through the list of issues.
@zoepage said:
> The idea is to create an UI to find issues with similar domain names and maybe even more. 

Maybe it's what it does first but what is the outcome.

@adamopenweb said:
> If there is a report for this domain that has a lot of activity, it's possible we have a common duplicate.

This is one possible outcome, but which is more on the triage side than the reporter side **currently**. The question is with the bag of data we have, how do we surface this information?


### About the workflow.
Let's assume that this is a logged-in user feature (!= power user). The information maybe would be of better quality in a second step.
1. Bug form report. Submit (anonymous or GitHub)
2. Next stage
    - Anonymous goes to the issue page. Nothing changed.
    - GitHub. Goes to a post-report page. Giving clues about possible previous matches and proposing to post a comment with the references to other issues. It gives more space for dealing with it. It removes some of the UI difficulties which comes with it. And maybe we can find strategies for the search/match.

## About the search API 😭 

@zoepage said:
> The shortest possible domain would be some thing like x.y, but we could check the first 2 chars for the www.string and fire the event starting at the 5th character then.

Yes that would be better! So I tried again this morning with "quoted string"

* ["facebook.com"](https://github.com/webcompat/web-bugs/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20%22facebook.com%22%20in%3Atitle)

Pain… then… The silly me though let's try to replace the `.` with a blank space.

* ["facebook com"](https://github.com/webcompat/web-bugs/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20%22facebook%20com%22%20in%3Atitle%20)

Oh Joy 🎉 this seems to be working.

* ["www facebook com"](https://github.com/webcompat/web-bugs/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20%22www%20facebook%20com%22%20in%3Atitle%20)

Let's try with ["docs google com"](https://github.com/webcompat/web-bugs/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20%22docs%20google%20com%22%20in%3Atitle%20)

😂  It removes a big big big 🤕 headache from even trying to create a prototype.

## Web site vs extension vs Site Reporter Menu

@zoepage said:
> why this feature would only be available on the site itself and not in the extension.

These are the same page. If we want to make a difference, we would need to create an http header into there, if it's not the case yet. @miketaylr do we still send the reporter header?


@zoepage said:
> So you'd dismiss this proposal completely and go with the localDB solution?

That sounded dramatic 🎭 As I said, we are thinking about it and sorry if it sounded negative, that was not the intent. I'm trying to explore all the hurdles and issues we can meet. I think the localDB solution is indeed a must in a very near future with its own pile of muddy issues. 😿 It's not fun.

I have the feeling we may have put the cart before the horses, because of the history of the issue and our own habits when using bugzilla. Maybe our feature is not "search similar domains", but more "user can do **this** with being exposed to similar domains." aka what are we trying to solve. So maybe these things:

At reporting stage:

* Find duplicate issues (specifically the original one)
  - to not report?
  - to leave an automatic comment with potential duplicates?

Maybe the first **prototype**, not something that would ever go to the production site but could be deployed on staging. Would be to just try to pull the search issues list when typing the domain and see if we get something meaningful. Not triggering any further action. We might not have to even do it on the site at first stage.

1. simple web page, showing the search results when typing in an input field. (somewhere on the web)
2. Simple search list on the form (on staging)
3. Next steps and so on.

Like this we don't develop a big machine, we learn from what we see, we try to tweak if we can improve the data quality, we learn what kind of actions we can do with this list or abandon if not fruitful.

That makes the goal less ambitious, but maybe more achievable through small iterations.

**Comment by miketaylr** [comment]
> I wonder how we could improve this. TO THINK.

I think generally the answer is to debounce or throttle (or both) function calls so they fire no more than N times per millisecond or event. We could experiment with this to get it right. And who knows, get crazy with dbs on the client or server.

> 😂 It removes a big big big 🤕 headache from even trying to create a prototype.

This is ridiculous haha. Good find.

> If we want to make a difference, we would need to create an http header into there, if it's not the case yet. @miketaylr do we still send the reporter header?

We send a GET param, `reported-with: foo` and anything that is not "web" is an addon initiated pageload.

> Maybe the first prototype, not something that would ever go to the production site but could be deployed on staging. Would be to just try to pull the search issues list when typing the domain and see if we get something meaningful.

This is interesting because it would potentially also be useful as a standalone tool for triaging, which is sort of what @adamopenweb was saying.

**Comment by MDTsai** [comment]
Some ideas about searching issues:
First, current issue title is not helpful, "Site is not usable" means different things. Sometimes description is too long but helpful to know what's happen.
Second, as my experience to figure out "duplicate" issues, we have issues with same root cause (radio/checkbox in android, text not align well, default font, etc), but they are from different domain. A better search would be helpful to reduce number of issues filed. That also means less time spend on triage same root cause issues.

**Comment by zoepage** [comment]
Closing as we move the discussion to #1846
